File: Native/src/main/java/com/example/mindflip_native/views/FlashcardsScreen.kt
package com.example.mindflip_native.views

import android.R
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material.icons.outlined.FileCopy
import androidx.compose.material.icons.outlined.Folder
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.mindflip_native.models.Flashcard
import com.example.mindflip_native.viewModels.MainViewModel

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FlashcardsScreen(
    viewModel: MainViewModel,
    categoryId: Int,
    onAddFlashcard: () -> Unit,
    onBack: () -> Unit,
    onEditFlashcard: (Int) -> Unit,
    onPractice: () -> Unit
) {
    val flashcards by viewModel.flashcards.observeAsState(emptyList())
    var flashcardToDelete by remember { mutableStateOf<Flashcard?>(null) }
    val categoryName = viewModel.categories.value?.find { it.localId == categoryId }?.name ?: "Category"

    Scaffold(
        topBar = {
            Column {
                TopAppBar(
                    title = { Text(categoryName, color = Color.Black, fontSize = 22.sp) },
                    navigationIcon = {
                        IconButton(onClick = onBack) {
                            Icon(
                                Icons.Default.ArrowBack,
                                contentDescription = "Back",
                                tint = Color.Black
                            )
                        }
                    },
                    actions = {
                        Button(
                            onClick = onPractice,
                            colors = ButtonDefaults.buttonColors(containerColor = Color(0xFF2196F3)),
                            shape = RoundedCornerShape(10.dp),
                            modifier = Modifier
                                .height(40.dp)
                        ) {
                            Text("Practice", color = Color.White)
                        }
                    },
                    colors = TopAppBarDefaults.centerAlignedTopAppBarColors(
                        containerColor = Color(0xFFF0F0F0)
                    )
                )
                Divider(color = Color.Black.copy(alpha = 0.07f), thickness = 1.dp)
            }
        },
        containerColor = Color(0xFFF0F0F0),
        bottomBar = {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .background(Color.Transparent),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Spacer(modifier = Modifier.height(8.dp))
                Button(
                    onClick = onAddFlashcard,
                    modifier = Modifier
                        .fillMaxWidth(0.7f)
                        .height(60.dp),
                    colors = ButtonDefaults.buttonColors(containerColor = Color(0xFFFF8C00)),
                    shape = RoundedCornerShape(60.dp)
                ) {
                    Text("+ Add Flashcard", color = Color.White, fontSize = 22.sp)
                }
                Spacer(modifier = Modifier.height(30.dp))
            }
        },
    ) { paddingValues ->
        LazyColumn(
            modifier = Modifier
                .fillMaxSize()
                .padding(top = paddingValues.calculateTopPadding(), bottom = paddingValues.calculateBottomPadding())
                .padding(horizontal = 16.dp)
        ) {
            items(flashcards) { flashcard ->
                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 8.dp),
                    colors = CardDefaults.cardColors(containerColor = Color.White),
                    elevation = CardDefaults.cardElevation(defaultElevation = 6.dp),
                    shape = RoundedCornerShape(8.dp)
                ) {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(12.dp),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.SpaceBetween
                    ) {
                        // Left square with folder icon
                        Box(
                            modifier = Modifier
                                .size(50.dp)
                                .clip(RoundedCornerShape(4.dp))
                                .background(Color(0xFFF0F0F0)),
                            contentAlignment = Alignment.Center
                        ) {
                            Icon(
                                imageVector = Icons.Outlined.FileCopy,
                                contentDescription = "FileCopy",
                                tint = Color.Black
                            )
                        }

                        Spacer(modifier = Modifier.width(12.dp))

                        Column(modifier = Modifier.weight(1f)) {
                            Text(
                                text = flashcard.question,
                                fontSize = 16.sp,
                                color = Color.Black,
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis
                            )

                            val diffColor = when (flashcard.difficulty.name) {
                                "EASY" -> Color(0xFF4CAF50)
                                "MEDIUM" -> Color(0xFFFFA500)
                                "HARD" -> Color(0xFFF44336)
                                else -> Color.Black
                            }

                            Text(
                                text = flashcard.difficulty.name.lowercase()
                                    .replaceFirstChar { it.uppercase() },
                                fontSize = 14.sp,
                                color = diffColor
                            )
                        }

                        Row {
                            IconButton(
                                onClick = { onEditFlashcard(flashcard.localId) },
                                modifier = Modifier.size(36.dp)
                            ) {
                                Icon(Icons.Default.Edit, contentDescription = "Edit Flashcard", modifier = Modifier.size(20.dp))
                            }
                            IconButton(
                                onClick = { flashcardToDelete = flashcard },
                                modifier = Modifier.size(36.dp)
                            ) {
                                Icon(Icons.Default.Delete, contentDescription = "Delete Flashcard", modifier = Modifier.size(20.dp))
                            }
                        }
                    }
                }
            }
        }

        // Delete dialog
        flashcardToDelete?.let { flashcard ->
            AlertDialog(
                onDismissRequest = { flashcardToDelete = null },
                title = {
                    Text(
                        "Do you really want to delete the card?",
                        textAlign = TextAlign.Center,
                        modifier = Modifier.fillMaxWidth()
                    )
                },
                confirmButton = {
                    Button(
                        onClick = {
                            viewModel.deleteFlashcard(flashcardToDelete!!.localId, categoryId)
                            flashcardToDelete = null
                        },
                        colors = ButtonDefaults.buttonColors(containerColor = Color(0xFFFF8C00)),
                        modifier = Modifier
                            .padding(horizontal = 8.dp),

                    ) {
                        Text("Yes")
                    }
                },
                dismissButton = {
                    Button(
                        onClick = { flashcardToDelete = null },
                        colors = ButtonDefaults.buttonColors(containerColor = Color(0xFFD3D3D3)),
                        modifier = Modifier
                            .padding(horizontal = 8.dp)
                    ) {
                        Text("No", color = Color.Black)
                    }
                },
                containerColor = Color(0xFFF0F0F0),
                modifier = Modifier.shadow(8.dp, shape = RoundedCornerShape(16.dp))
            )

        }
    }
}

File: non_native/lib/views/flashcards_screen.dart

import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../view_models/main_view_model.dart';
import '../models/flashcard.dart';

class FlashcardsScreen extends StatelessWidget {
  final int categoryId;
  final VoidCallback onBack;
  final VoidCallback onAddFlashcard;
  final ValueChanged<int> onEditFlashcard;
  final VoidCallback onPractice;

  const FlashcardsScreen({
    super.key,
    required this.categoryId,
    required this.onBack,
    required this.onAddFlashcard,
    required this.onEditFlashcard,
    required this.onPractice,
  });

  void _showDeleteDialog(BuildContext context, Flashcard flashcard, MainViewModel viewModel) {
    showDialog(
      context: context,
      builder: (BuildContext dialogContext) {
        return AlertDialog(
          title: const Text("Do you really want to delete the card?", textAlign: TextAlign.center),
          actionsAlignment: MainAxisAlignment.center,
          backgroundColor: const Color(0xFFF0F0F0),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16.0)),
          actions: <Widget>[
            ElevatedButton(
              onPressed: () {
                Navigator.of(dialogContext).pop();
                viewModel.deleteFlashcard(flashcard.localId, flashcard.categoryId);
              },
              style: ElevatedButton.styleFrom(
                backgroundColor: const Color(0xFFFF8C00),
                padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 10),
              ),
              child: const Text("Yes", style: TextStyle(color: Colors.white)),
            ),
            ElevatedButton(
              onPressed: () {
                Navigator.of(dialogContext).pop();
              },
              style: ElevatedButton.styleFrom(
                backgroundColor: const Color(0xFFD3D3D3),
                padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 10),
              ),
              child: const Text("No", style: TextStyle(color: Colors.black)),
            ),
          ],
        );
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    final viewModel = context.watch<MainViewModel>();
    final flashcards = viewModel.flashcards;
    final categoryName = viewModel.getCategoryName(categoryId);

    return Scaffold(
      appBar: AppBar(
        title: Text(categoryName, style: const TextStyle(color: Colors.black, fontSize: 22.0)),
        backgroundColor: const Color(0xFFF0F0F0),
        leading: IconButton(
          icon: const Icon(Icons.arrow_back, color: Colors.black),
          onPressed: onBack, 
        ),
        actions: [
          Padding(
            padding: const EdgeInsets.only(right: 12.0),
            child: ElevatedButton(
              onPressed: onPractice, 
              style: ElevatedButton.styleFrom(
                backgroundColor: const Color(0xFF2196F3),
                shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(10.0)),
                minimumSize: const Size(0, 40.0),
              ),
              child: const Text("Practice", style: TextStyle(color: Colors.white)),
            ),
          ),
        ],
        bottom: PreferredSize(
          preferredSize: const Size.fromHeight(1.0),
          child: Divider(color: Colors.black.withOpacity(0.07), thickness: 1.0),
        ),
      ),
      body: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 16.0),
        child: ListView.builder(
          itemCount: flashcards.length,
          itemBuilder: (context, index) {
            final flashcard = flashcards[index];
            Color diffColor;

            switch (flashcard.difficulty.name) {
              case 'easy':
                diffColor = const Color(0xFF4CAF50);
                break;
              case 'medium':
                diffColor = const Color(0xFFFFA500);
                break;
              case 'hard':
                diffColor = const Color(0xFFF44336);
                break;
              default:
                diffColor = Colors.black;
            }

            return Card(
              elevation: 6.0,
              shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8.0)),
              margin: const EdgeInsets.symmetric(vertical: 8.0),
              child: Padding(
                padding: const EdgeInsets.all(12.0),
                child: Row(
                  children: [
                    
                    Container(
                      width: 50.0,
                      height: 50.0,
                      decoration: BoxDecoration(
                        color: const Color(0xFFF0F0F0),
                        borderRadius: BorderRadius.circular(4.0),
                      ),
                      child: const Icon(Icons.copy_outlined, color: Colors.black),
                    ),
                    const SizedBox(width: 12.0),
                    
                    Expanded(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Text(
                            flashcard.question,
                            style: const TextStyle(fontSize: 16.0, color: Colors.black),
                            maxLines: 1,
                            overflow: TextOverflow.ellipsis,
                          ),
                          Text(
                            
                            '${flashcard.difficulty.name[0].toUpperCase()}${flashcard.difficulty.name.substring(1)}',
                            style: TextStyle(fontSize: 14.0, color: diffColor),
                          ),
                        ],
                      ),
                    ),
                    
                    Row(
                      children: [
                        IconButton(
                          icon: const Icon(Icons.edit, size: 20.0),
                          onPressed: () => onEditFlashcard(flashcard.localId), 
                        ),
                        IconButton(
                          icon: const Icon(Icons.delete, size: 20.0),
                          onPressed: () => _showDeleteDialog(context, flashcard, viewModel),
                        ),
                      ],
                    ),
                  ],
                ),
              ),
            );
          },
        ),
      ),
      bottomNavigationBar: Padding(
        padding: const EdgeInsets.only(left: 16.0, right: 16.0, bottom: 30.0),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            ElevatedButton(
              onPressed: onAddFlashcard, 
              style: ElevatedButton.styleFrom(
                backgroundColor: const Color(0xFFFF8C00),
                shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(60.0)),
                minimumSize: const Size(double.infinity, 60.0),
              ),
              child: const Text("+ Add Flashcard", style: TextStyle(color: Colors.white, fontSize: 22.0)),
            ),
          ],
        ),
      ),
    );
  }
}